### PR TITLE
Conveyors

### PR DESCRIPTION
**Pull request by ThatDan123**
Adds Conveyor Belts and Conveyor Belt Switches

-Takes 5 metal to build either.

-Use multitool to link conveyors to switches, can click multiple conveyors before linking to switch.

-Use wrench to change direction of conveyor belt.

-Use crowbar to deconstruct conveyor belt or switch.

-Use screwdriver on switch to clear it.

-Use screwdriver on conveyor to invert it.

-Outpost has cargo conveyors mapped.

-Conveyors move items, objects and players only if the destination tile is passable.

Thanks Gilles and ReiderJack for your help!

~Conveyor Belts added, currently adding switches~

~Also construction still needs to be done.~

~Only issue left is that on headless the client isnt syncing ConveryorStatus for whatever reason.~

STILL ISSUES

-Animation not playing on headless when you manually construct/spawn in conveyors, works on mapped conveyors.

-In particular directions you get teleported/jumped quicker, not sure whats the cause of this.

- [x] Code is sufficiently commented
- [x] Code is indented with tabs and not spaces
- [x] The PR does not include any unnecessary .meta, .prefab or **.unity (scene) changes**
- [x] The PR does not bring up any new compile errors
- [x] The PR has been tested in editor
- [x] The PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)
- [x] The PR has been tested with round restarts.

### CODE REVIEW

File: UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+			{
+				spriteHandler.ChangeSprite(2);
+			}
+			else
+			{
+				spriteHandler.ChangeSprite(0);
+			}
+
+			if (CurrentDirection == ConveyorDirection.Up)
+			{
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs]: Use `switch` statement instead all those `if`s
ReviewSwitch

File: UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+		Logger.Log("Current status "+ CurrentStatus);
+	}
+
+	private void ChangeAnimation()
+	{
+		//if (CurrentDirection != LastDirection || CurrentStatus != LastStatus)
+		//{
+			if (CurrentStatus == ConveyorStatus.Off)
+			{
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs]: Use `switch`, put `spriteHandler.ChangeSprite(0);` under `default:`

File: UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+	private void ChangeDirection()
+	{
+		if (ConnectedSwitch == null) return;
+
+		LastStatus = CurrentStatus;
+
+		Logger.Log("Connected Switch: "+ ConnectedSwitch.CurrentState);
+
+		if (ConnectedSwitch.CurrentState == 0)
+		{
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs:128]: Put this in the method:
ChangeSwitch

File: UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	public ConveyorBelt[] conveyorBelts;
+
+	public Sprite Forward;
+	public Sprite Backward;
+	public Sprite Off;
+
+	private Sprite LastSprite;
+
+	public int CurrentState = 1; // 0 Backwards, 1 Off, 2 Forwards
+
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs:20]: Make it a `enum`, don't use `int `for states.

File: UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		if (!Validations.IsTarget(gameObject, interaction)) return false;
+
+		return true;
+	}
+
+	public void ServerPerformInteraction(HandApply interaction)
+	{
+		if (spriteRenderer.sprite == Forward)
+		{
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs]: Compare to `CurrentState ` not sprites. Also use `switch `here.

File: UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs (hunk 4 above)
Comment (ReiderJack) [reply UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs:20]: You can use it for Conveyor and Switch
ConveyorStatus

File: UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ConveyorBeltSwitch : MonoBehaviour, ICheckedInteractable<HandApply>
+{
+	public SpriteRenderer spriteRenderer;
+
+	public ConveyorBelt[] conveyorBelts;
+
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs:13]: You will be able to set up all conveyors in editor, but in game `array `won't work. So you will probably need a `List<ConveyorBelt>`, if you are gonna implement construction and connection to new switches.

File: UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+
+	//void Start()
+	//{
+	//	registerTile = GetComponent<RegisterTile>();
+
+	//	CurrentDirection = MappedDirection;
+	//	CurrentStatus = MappedStatus;
+	//	LastDirection = CurrentDirection;
+	//	LastStatus = CurrentStatus;
+
```
Comment (ReiderJack) [inline UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs:95]: Do not commit zombie code

Comment (corp-0) [reply UnityProject/Assets/Scripts/Construction/ConveyorBelt.cs:95]: It is a draft pr

**Comment by corp-0** [comment]
I did a PR in your branch with a suggested refactor

File: UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs (hunk 6 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Construction/ConveyorBeltSwitch.cs:13]: Yeah thx, ive changed it now.

**Comment by ThatDan123** [comment]
Last issue to solve: 

-Animation not playing on headless when you manually construct/spawn in conveyors, however works on mapped conveyors.

**Comment by DooblyNoobly** [comment]
Nice work, I'll merge this progress as there is scene changes involved.

**Pull request merged**